engine/src/org/pentaho/di/trans/steps/datagrid/DataGrid.java

```diff
@@ -84,16 +84,19 @@ public boolean processRow( StepMetaInterface smi, StepDataInterface sdi ) throws
     List<String> outputLine = meta.getDataLines().get( data.linesWritten );
 
     for ( int i = 0; i < data.outputRowMeta.size(); i++ ) {
-      if ( meta.isSetEmptyString()[i] ) {
+      if ( meta.isSetEmptyString()[ i ] ) {
         // Set empty string
-        outputRowData[i] = StringUtil.EMPTY_STRING;
+        outputRowData[ i ] = StringUtil.EMPTY_STRING;
       } else {
 
         ValueMetaInterface valueMeta = data.outputRowMeta.getValueMeta( i );
         ValueMetaInterface convertMeta = data.convertMeta.getValueMeta( i );
         String valueData = outputLine.get( i );
 
-        outputRowData[i] = valueMeta.convertDataFromString( valueData, convertMeta, null, null, 0 );
+        if ( valueData != null && valueMeta.isNull( valueData ) ) {
```

Comment:
ValueMetaBase.isNull() is already designed to check whether valueData is null.  Seems odd to pre-check it before calling the method that is specifically designed to test the nullness of objects.